File: app/src/main/java/com/kjjejones42/audiobookplayer/MediaItem.java
package com.kjjejones42.audiobookplayer;

import android.content.Context;
import android.content.res.AssetFileDescriptor;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.media.MediaMetadataRetriever;
import android.net.Uri;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.io.ByteArrayInputStream;
import java.io.FileDescriptor;
import java.io.Serializable;

public class MediaItem implements Serializable, Comparable<MediaItem> {

    private final String uri;
    private final String displayName;
    private final long duration;
    private transient MediaMetadataRetriever mmr;

    @Nullable
    private MediaMetadataRetriever getMMR(Context context) {
        try {
            if (mmr == null) {
                AssetFileDescriptor assetFileDescriptor = context.getContentResolver().openAssetFileDescriptor(getUri(), "r");
                if (assetFileDescriptor != null) {
                    FileDescriptor fileDescriptor = assetFileDescriptor.getFileDescriptor();
                    mmr = new MediaMetadataRetriever();
                    mmr.setDataSource(fileDescriptor);
                }
            }
            return mmr;
        } catch (Exception e) {
            return null;
        }
    }

    public Bitmap getEmbeddedPicture(Context context) {
        Bitmap result = null;
        try {
            mmr = getMMR(context);
            if (mmr != null) {
                ByteArrayInputStream bis = new ByteArrayInputStream(mmr.getEmbeddedPicture());
                result = BitmapFactory.decodeStream(bis);
                bis.close();
                mmr.close();
            }
        } catch (Exception ignored) {}
        return result;
    }

    public MediaItem(Uri documentUri, String displayName, long duration) {
        this.uri = documentUri.toString();
        this.displayName = displayName;
        this.duration = duration;
    }

    public Uri getUri() {
        return Uri.parse(uri);
    }
    long getDuration() {
        return duration;
    }

    @NonNull
    @Override
    public String toString() {
        int index = displayName.lastIndexOf('.');
        if (index == -1) {
            return displayName;
        }
        return displayName.substring(0, index);
    }
    @Override
    public int compareTo(MediaItem o) {
        return uri.compareTo(o.uri);
    }
}

File: app/src/main/java/com/kjjejones42/audiobookplayer/MediaItem.kt
package com.kjjejones42.audiobookplayer

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.media.MediaMetadataRetriever
import android.net.Uri
import androidx.core.net.toUri
import com.kjjejones42.audiobookplayer.database.DataConverter
import java.io.ByteArrayInputStream
import java.io.Serializable
import java.util.Objects

@kotlinx.serialization.Serializable(with = DataConverter::class)
class MediaItem(uri: Uri, private val displayName: String, val fileName: String, @JvmField val duration: Long) :
Serializable, Comparable<MediaItem?> {

    private var _uri = uri.toString()
    var uri: Uri
        get() = _uri.toUri()
        set(value) {
            _uri = value.toString()
        }

    private fun getMMR(context: Context): MediaMetadataRetriever? {
        return try {
            context.contentResolver.openAssetFileDescriptor(uri, "r")?.use {
                val fileDescriptor = it.fileDescriptor
                MediaMetadataRetriever().also {it.setDataSource(fileDescriptor)}
            }
        } catch (e: Exception) {
            throw RuntimeException(e)
        }
    }

    fun getEmbeddedPicture(context: Context): Bitmap? {
        return try {
            getMMR(context)?.use {
                ByteArrayInputStream(it.embeddedPicture).use {
                    BitmapFactory.decodeStream(it)
                }
            }
        } catch (_: Exception) { null }
    }


    override fun toString(): String {
        val index = displayName.lastIndexOf('.')
        if (index == -1) {
            return displayName
        }
        return displayName.take(index)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        return uri == (other as MediaItem).uri
    }

    override fun hashCode(): Int {
        return Objects.hash(uri, displayName, duration)
    }

    override fun compareTo(other: MediaItem?): Int {
        if (other == null) return 0
        return fileName.compareTo(other.fileName)
    }
}
